apps/extension/src/extension/Session.ts

```diff
@@ -0,0 +1,120 @@
+import browser, { Runtime } from "webextension-polyfill";
+import { LockKeyRingMsg } from "background/keyring";
+import { Ports } from "router";
+import { ExtensionRequester } from "./ExtensionRequester";
+
+export enum SessionMsgType {
+  Connect = "connect",
+  Disconnect = "disconnect",
+  Message = "message",
+}
+
+export type SessionMsg = {
+  msg: string;
+  type: SessionMsgType;
+};
+
+const SESSION_PORT = "session-port";
+// Maximum length of inactivity
+const SESSION_DURATION = 4 * 60 * 1000;
+
+type Timeout = ReturnType<typeof setTimeout>;
+// Connect to SESSION_PORT
+const initPort = (): Runtime.Port =>
+  browser.runtime.connect({
+    name: SESSION_PORT,
+  });
+
+const clearSessionTimeout = (timeout?: Timeout): void => {
+  if (timeout) {
+    clearTimeout(timeout);
+  }
+};
+
+/**
+ * Maintain an open port to the background until a certain limit has been reached.
+ */
+export class Session {
+  protected port: Runtime.Port | undefined;
+  private _requester: ExtensionRequester | undefined;
+  private _timestamp: number = Date.now();
+  private _timeout: Timeout | undefined;
+
+  public start(requester?: ExtensionRequester): Runtime.Port {
+    this._timestamp = Date.now();
+    // Restart port:
+    this.port?.disconnect();
+    this.port = initPort();
+    this.port.postMessage({
+      msg: "Establishing port to background",
+      type: SessionMsgType.Connect,
+    });
+    if (requester) {
+      this._requester = requester;
+    }
+
+    this.port.onMessage.addListener(this.handleOnMessage);
+    this.port.onDisconnect.addListener(this.handleOnDisconnect);
+
+    this._timeout = setTimeout(() => {
+      const duration = Date.now() - this._timestamp;
+      if (duration < SESSION_DURATION) {
```

Comment:
Can this condition ever be true (and cause us to update the session)? as there are always a few ms from where we set `this._timestamp` to where we call `setTimeout`.

The purpose of this method is to start a session and at the end of the timeout automatically send `SessionMsgType.Disconnect` to terminate the session or it should also be keeping it alive?

Reply:
This has been removed.